File: embeddings/index-library.py
#!/usr/bin/env python3
"""
Index tracks from Music Assistant API into database.
Run locally: python embeddings/index-library.py
"""

import os
import requests
import psycopg2


def main():
    DATABASE_URL = os.getenv("DATABASE_URL")
    if not DATABASE_URL:
        raise ValueError("DATABASE_URL not set")

    HOST = os.getenv("HOST")
    TOKEN = os.getenv("TOKEN")
    CONFIG_ID = os.getenv("CONFIG_ID")

    if not HOST or not TOKEN or not CONFIG_ID:
        raise ValueError("HOST, TOKEN, and CONFIG_ID must be set")

    print(f"Fetching tracks from {HOST}...")

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {TOKEN}",
    }

    payload = {
        "limit": "100000",
        "library_only": "true",
        "config_entry_id": CONFIG_ID,
        "name": "",
        "media_type": "track",
    }

    response = requests.post(
        f"{HOST}/api/services/music_assistant/search?return_response",
        headers=headers,
        json=payload,
    )
    response.raise_for_status()

    data = response.json()
    tracks_data = data["service_response"]["tracks"]

    print(f"Found {len(tracks_data)} tracks from API")

    conn = psycopg2.connect(DATABASE_URL)
    cur = conn.cursor()

    inserted = 0

    for track in tracks_data:
        track_uri = track["uri"]
        track_name = track["name"]
        album_name = track["album"]["name"]
        artists_list = [a["name"] for a in track["artists"]]

        cur.execute(
            """
            INSERT INTO track (uri, name, artist, album)
            VALUES (%s, %s, %s, %s)
            ON CONFLICT (uri) DO NOTHING
            """,
            (track_uri, track_name, artists_list, album_name),
        )
        if cur.rowcount > 0:
            inserted += 1

    conn.commit()
    cur.close()
    conn.close()

    print(f"\nComplete: {inserted} tracks inserted")


if __name__ == "__main__":
    main()

File: web/src/lib/server/index-library.ts
import { sql } from 'drizzle-orm';
import { db } from './db';
import { trackTable } from './schema';

const BATCH_SIZE = 500;
const CHUNK_SIZE = 150;

interface Artist {
	media_type: string;
	uri: string;
	name: string;
	version: string;
	image?: string;
}

interface Album {
	media_type: string;
	uri: string;
	name: string;
	version: string;
	image: string;
	artists: Artist[];
}

interface Track {
	media_type: string;
	uri: string;
	name: string;
	version: string;
	image: string;
	artists: Artist[];
	album: Album;
}

function normalizeArtistName(name: string): string {
	return name
		.replace(/\*/g, '')
		.replace(/_/g, '')
		.normalize('NFD')
		.replace(/[\u0300-\u036f]/g, '')
		.replace(/\sft\.\s/gi, ' feat. ')
		.replace(/\sft\s/gi, ' feat. ')
		.replace(/\sfeat\.\s/gi, ' feat. ')
		.replace(/\sfeat\s/gi, ' feat. ')
		.replace(/\sfeaturing\s/gi, ' feat. ')
		.replace(/with\s/gi, ' w/ ')
		.replace(/&/g, ' & ')
		.replace(/\s+/g, ' ')
		.trim();
}

function splitArtists(artistName: string): string[] {
	const normalized = normalizeArtistName(artistName);
	const parts = normalized.split(/ feat\.? | vs\.? | & |, | w\/| x | \/ | \+ | duet /);
	return parts.map((p) => p.trim()).filter((p) => p.length > 0);
}

/**
 * Keep only the imageproxy path from a Music Assistant cover URL.
 *
 * The upstream host is a private LAN address, so storing the full URL would
 * bake deployment-specific detail into the database and break whenever the
 * host changes. The path is stable; the host is re-applied at render time.
 */
function imageProxyPath(image: string | undefined): string | null {
	if (!image) return null;
	try {
		const url = new URL(image);
		if (!url.pathname.startsWith('/imageproxy/')) return null;
		return url.pathname;
	} catch {
		return null;
	}
}

export async function indexLibrary(): Promise<number> {
	const env = process.env;

	const authHeaders = new Headers();
	authHeaders.append('Content-Type', 'application/json');
	authHeaders.append('Authorization', 'Bearer ' + env.TOKEN);

	let offset = 0;
	let allTracks: Track[] = [];

	console.log('Fetching library tracks from external service using pagination...');

	while (true) {
		const raw = JSON.stringify({
			config_entry_id: env.CONFIG_ID,
			media_type: 'track',
			limit: BATCH_SIZE,
			offset,
			order_by: 'sort_name'
		});

		console.log(`Fetching tracks batch: offset ${offset}, limit ${BATCH_SIZE}`);

		const res = await fetch(
			env.HA_HOST + '/api/services/music_assistant/get_library?return_response',
			{ method: 'POST', headers: authHeaders, body: raw, redirect: 'follow' }
		);

		const text = await res.text();

		if (res.status !== 200) {
			console.error('Response text:', text);
			throw new Error(`HTTP ${res.status}: ${text}`);
		}

		let data: any;
		try {
			data = JSON.parse(text);
		} catch (error) {
			console.error('Failed to parse JSON response:', error);
			console.error('Response text:', text);
			throw error;
		}

		const batch = data.service_response.items;

		if (!batch || batch.length === 0) {
			console.log('No more tracks to fetch.');
			break;
		}

		allTracks.push(...batch);
		console.log(`Fetched ${batch.length} tracks. Total: ${allTracks.length}`);

		if (batch.length < BATCH_SIZE) {
			console.log('Reached end of library tracks.');
			break;
		}

		offset += BATCH_SIZE;
	}

	const tracks = allTracks.map((track: Track) => {
		const artistNames = track.artists.flatMap((artist: Artist) => splitArtists(artist.name));
		return {
			name: track.name,
			uri: track.uri,
			artist: artistNames,
			album: track.album.name,
			albumImage: imageProxyPath(track.album.image)
		} satisfies typeof trackTable.$inferInsert;
	});

	console.log(`Starting batch insert/update for ${tracks.length} tracks...`);

	for (let i = 0; i < tracks.length; i += CHUNK_SIZE) {
		const chunk = tracks.slice(i, i + CHUNK_SIZE);
		try {
			await db
				.insert(trackTable)
				.values(chunk)
				.onConflictDoUpdate({
					target: trackTable.uri,
					set: {
						artist: sql`EXCLUDED.artist`,
						album: sql`EXCLUDED.album`,
						albumImage: sql`EXCLUDED.album_image`,
						updatedAt: sql`CURRENT_TIMESTAMP`
					}
				});
			console.log(
				`Success: Chunk ${i / CHUNK_SIZE + 1} / ${Math.ceil(tracks.length / CHUNK_SIZE)}`
			);
		} catch (error) {
			console.error(`Error in chunk starting at ${i}:`, error);
		}
	}

	console.log('Finished processing tracks into the database.');
	return tracks.length;
}
